worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java

```diff
@@ -17,615 +17,149 @@
-    }
+    return tierWriterProxy.needHardSplitForMemoryFile();
   }
 
   /** assume data size is less than chunk capacity */
   public void write(ByteBuf data) throws IOException {
-    if (closed) {
-      String msg = getFileAlreadyClosedMsg();
-      logger.warn(msg);
-      throw new AlreadyClosedException(msg);
-    }
-
-    if (notifier.hasException()) {
-      if (s3MultipartUploadHandler != null) {
-        logger.warn("Abort s3 multipart upload for {}", diskFileInfo.getFilePath());
-        s3MultipartUploadHandler.complete();
-        s3MultipartUploadHandler.close();
-      }
-      return;
-    }
-
-    int mapId = 0;
-    if (rangeReadFilter) {
-      byte[] header = new byte[4];
-      data.markReaderIndex();
-      data.readBytes(header);
-      data.resetReaderIndex();
-      mapId = Platform.getInt(header, Platform.BYTE_ARRAY_OFFSET);
-    }
+    tierWriterProxy.write(data);
+    tierWriterProxy.decrementPendingWriters();
+  }
 
-    final int numBytes = data.readableBytes();
-
-    synchronized (flushLock) {
-      if (closed) {
-        String msg = getFileAlreadyClosedMsg();
-        logger.warn(msg);
-        throw new AlreadyClosedException(msg);
-      }
-      if (rangeReadFilter) {
-        mapIdBitMap.add(mapId);
-      }
-      int flushBufferReadableBytes = flushBuffer.readableBytes();
-      if (!isMemoryShuffleFile.get()) {
-        if (flushBufferReadableBytes != 0
-            && flushBufferReadableBytes + numBytes >= flusherBufferSize) {
-          flush(false, false);
-        }
-      } else {
-        if (flushBufferReadableBytes > memoryFileStorageMaxFileSize
-            && storageManager.localOrDfsStorageAvailable()) {
-          logger.debug(
-              "{} Evict, memory buffer is  {}",
-              writerContext.getPartitionLocation().getFileName(),
-              flushBufferReadableBytes);
-          evict(false);
-        }
-      }
-
-      // update the disk buffer or memory file storage after evict
-      if (isMemoryShuffleFile.get()) {
-        MemoryManager.instance().incrementMemoryFileStorage(numBytes);
-      } else {
-        MemoryManager.instance().incrementDiskBuffer(numBytes);
-        if (userCongestionControlContext != null) {
-          userCongestionControlContext.updateProduceBytes(numBytes);
-        }
-      }
-
-      data.retain();
-      try {
-        flushBuffer.addComponent(true, data);
-      } catch (OutOfMemoryError oom) {
-        data.release();
-        if (isMemoryShuffleFile.get()) {
-          MemoryManager.instance().releaseMemoryFileStorage(numBytes);
-        } else {
-          MemoryManager.instance().releaseDiskBuffer(numBytes);
-        }
-        throw oom;
-      }
-      if (isMemoryShuffleFile.get()) {
-        memoryFileInfo.updateBytesFlushed(numBytes);
-      }
+  public RoaringBitmap getMapIdBitMap() {
+    if (tierWriterProxy.currentTierWriter().metaHandler() instanceof MapPartitionMetaHandler) {
```

Comment:
MapPartitionMetaHandler should be consistent when PartitionDataWriter initialized?